near-plugins-derive/src/upgradable.rs

```diff
@@ -21,19 +22,29 @@ pub fn derive_upgradable(input: TokenStream) -> TokenStream {
         .code_storage_key
         .unwrap_or_else(|| "__CODE__".to_string());
 
+    let allowed_timestamp_storage_key = opts
+        .allowed_timestamp_storage_key
+        .unwrap_or_else(|| "__ALLOWED_TIMESTAMP__".to_string());
+
     let output = quote! {
         #[near_bindgen]
         impl Upgradable for #ident {
-            fn up_storage_key(&self) -> Vec<u8>{
+            fn up_storage_key(&self) -> Vec<u8> {
                 (#code_storage_key).as_bytes().to_vec()
             }
 
+            fn up_allowed_timestamp_storage_key(&self) -> Vec<u8> {
+                (#allowed_timestamp_storage_key).as_bytes().to_vec()
+            }
+
             #[#cratename::only(owner)]
-            fn up_stage_code(&mut self, #[serializer(borsh)] code: Vec<u8>) {
+            fn up_stage_code(&mut self, #[serializer(borsh)] code: Vec<u8>,  #[serializer(borsh)] delay_timestamp: u64) {
```

Comment:
Orthogonal to this PR. I wonder if we should avoid borsh here and use json, passing the code in base64. It will be more inefficient, but better for readability.